mantle/kola/README-kola-ext.md

```diff
@@ -10,31 +13,83 @@ repositories, and allow these projects to target Fedora CoreOS
+The `--exttest` (`-E`) argument to `kola run` one way to accomplish this; you
+provide the path to an upstream project git repository.  Tests will be found
+in the `tests/kola` directory.  If this project contains binaries that require
+building, it is assumed that `make` (or equivalent) has already been invoked.
+
+The `tests/kola` directory will be traversed recursively to find tests.
+
+The core idea is to express a test as a single binary (plus an optional
+directory of dependencies named `deps`) that will be uploaded to a CoreOS
+system and run as a systemd unit, along with an optional Ignition config
+named `config.ign`.
 
-The `kola run-ext-bin` is one way to accomplish this.  The
-core idea is to express your test suite as a binary (plus an optional
-directory of dependencies) that will be uploaded to a CoreOS
-system and run as a systemd unit.
+Concretely then, an external test directory can have the following content:
 
-Currently this systemd unit runs with full privileges - tests
+- `config.ign` (optional): Ignition config provided
+- `kola.json` (optional): JSON file described below
+- `deps` (optional): Directory (or symlink to dir): Will be uploaded and available as `${KOLA_EXT_DATA}`
+- one or more executables: Each executable is its own test, run independently
+  with the Ignition config and/or dependencies provided.
+
+In the case of a test directory with a single executable, the kola test name will be
+`ext.<projname>.<subdirectory>`.  Otherwise, the test will be named `ext.<projname>.<subdirectory>.<executable>`.
+
+Currently the test systemd unit runs with full privileges - tests
 are assumed to be (potentially) destructive and a general assumption
-is tests are run in easily disposable virtual machines.
+is tests are run in easily disposable virtual machines.  A future
+enhancement will support nondestructive tests.
 
 A best practice for doing this is to write your tests in a language
 that compiles to a single binary - Rust and Go for example, but
 there exist for example tools like [PyInstaller](https://realpython.com/pyinstaller-python/#pyinstaller)
-too.
+too.  This way you can usually avoid the need for a `deps` directory.
 
 This mechanism is suitable for testing most userspace components
 of CoreOS; for example, one can have the binary drive a container runtime.
 
-An important feature of `run-ext-bin` is support for rebooting the host system.
-This allows one to easily test OS updates.  To do this, simply invoke the usual
-`reboot` - the test framework will monitor the target systemd unit
-and ignore the case where it exits with `SIGTERM`.
-
 A test is considered failed if the unit exits with any non-zero exit
 status or dies from any signal other than `SIGTERM`.
 
+Support for rebooting
+---
+
+An important feature of exttests is support for rebooting the host system.
+This allows one to easily test OS updates for example.  To do this, your
+test process should send `SIGTERM` to itself.  For example, in bash use:
```

Comment:
I'm a little unsure about the `SIGTERM` convention here. It feels a little generic/implicit. Can we have something more explicit for requesting a reboot?

Hmm, how about a designated exit code? E.g. service exits 77, kolet itself exits 77, and kola reboots.

Reply:
(BTW, I really like that we're tackling the reboot problem here; this is something that the rpm-ostree testsuite will benefit from too!)

Reply:
I went with an exit code at first but, if we want to support having the test just invoke `reboot`, it's going to get `SIGTERM`, so I think this approach is better.  It feels pretty explicit to me.

Reply:
Right, I get that that's what happens anyway during a reboot. It just feels odd to go from there and flip this around to assume that if that's what happened, then it must mean we're rebooting. Anyway, I'm cool rolling with this to start with.